course6/toyapp/displaylib/surface.cc

```diff
@@ -15,23 +15,42 @@ Surface::~Surface()
 		SDL_FreeSurface(mSurface);
 }
 
-void Surface::Clear(std::uint8_t r, std::uint8_t g, std::uint8_t b)
+void Surface::Clear(Rgb color)
 {
 	assert(mSurface);
 
 	for(int y = 0; y < mSurface->h; y++)
 		for(int x = 0; x < mSurface->w; x++)
-			SetPixel(x, y, r, g, b);
+			SetPixel(Pixel{x, y}, color);
 }
 
-void Surface::SetPixel(int x, int y, std::uint8_t r, std::uint8_t g, std::uint8_t b)
+void Surface::SetPixel(Pixel position, Rgb color)
 {
 	assert(mSurface);
 
-	std::uint32_t pixelValue = (std::uint32_t(r)<<16) + (std::uint32_t(g)<<8) + b;
+	std::uint32_t pixelValue = (std::uint32_t(color.r)<<16) + (std::uint32_t(color.g)<<8) + color.b;
 
-	std::uint32_t* pixelAddr = static_cast<std::uint32_t*>(mSurface->pixels) + (y*mSurface->pitch/4 + x);
+	auto offset = position.y*mSurface->pitch/4 + position.x;
+	std::uint32_t* pixelAddr = static_cast<std::uint32_t*>(mSurface->pixels) + offset;
 	*pixelAddr = pixelValue;
 }
 
+void Surface::DrawVertex(Vector3D vertex, Rgb color)
+{
+	auto point2D = vertex.project();
+	Pixel halfScreen{mSurface->w / 2, mSurface->h / 2};
+	Pixel pixel {
+		static_cast<int>(halfScreen.x + point2D.x * halfScreen.x),
```

Comment:
Am I right that this is to move cube to center and scale it up?

Reply:
yes, this changes the coordinates from physical pixels to (-1.0 - +1.0)